### PR TITLE
Tests empty labels and honor

### LINKED ISSUE #3841: Empty labels being returned from PromQL
https://github.com/prometheus/alertmanager/issues/1240 shows empty labels somehow coming out of storage and into alerts. I've seen this previously myself also, but haven't narrowed it down. I suspect an issue on the ingestion path somewhere.

### PR DESCRIPTION
**Pull request by krasi-georgiev**
Fixes #3841 #3845

### CODE REVIEW

File: scrape/scrape_test.go [hunk 1]
```diff
@@ -1138,6 +1138,117 @@ func TestTargetScrapeScrapeNotFound(t *testing.T) {
...
+				},
+			},
+			Expected: []labels.Label{
+				{
+					Name:  "n",
+					Value: "v",
+				},
+			},
+		}, {
+			Title: "Delete When label name is empty",
```
Comment (brian-brazil) [inline scrape/scrape_test.go]: You also need to test that an empty scraped label overrides a target label when honor labels is set

Comment (krasi-georgiev) [reply scrape/scrape_test.go]: Just added this check as part of the `Honour Label` test.

**Comment by krasi-georgiev** [comment]
If I make it broader isn't this more of an integration test? 

At the moment is seems that we have a mix of unit/integration tests , but might want to start thinking if we can separate these a bit. Especially if we would decide to start working with the CNCF CI.

**Comment by krasi-georgiev** [comment]
My logic here is that we want to test the functionality of this `unit` - `mutateSampleLabels`

Also I find in the past that too broader test require refactoring on every code change which I am hoping it can be avoided by smaller unit tests.

I can add one more broader test , but would prefer if we keep this one as well.
Would also be also curious to see which one turns out to be more useful in the long run.

**Comment by fabxc** [comment]
I tend to agree that this is probably fine to test as a unit test.
If the parser gets a unit test itself for the empty label stuff it should be fine.

File: scrape/scrape_test.go [hunk 2]
```diff
@@ -1138,6 +1138,125 @@ func TestTargetScrapeScrapeNotFound(t *testing.T) {
...
+func TestMutateSampleLabels(t *testing.T) {
+
+	tests := []struct {
+		Title      string
+		ScrapePool *scrapePool
+		Input      struct {
+			Lset   labels.Labels
+			Target *Target
+		}
+		Expected labels.Labels
```
Comment (fabxc) [inline scrape/scrape_test.go]: No realt need to export all those.

File: scrape/scrape_test.go [hunk 3]
```diff
@@ -1138,6 +1138,125 @@ func TestTargetScrapeScrapeNotFound(t *testing.T) {
...
+				Target: &Target{
+					labels: []labels.Label{
+						{
+							Name:  "n",
+							Value: "v0",
+						},
+						{
+							Name:  "n1",
+							Value: "v1",
+						},
```
Comment (fabxc) [inline scrape/scrape_test.go]: I'd inline those like `{Name: "n1", Value: "v1"}` so the lines don't bloat up as bad. I find it hard to read a single test unit otherwise.

**Comment by brian-brazil** [comment]
> If I make it broader isn't this more of an integration test?

Integration tests involve running other binaries, which we're not doing here. I want a test for the feature that if a user specifies an empty scrape label with honor labels enabled that it will result in that label not being ingested, and this test does not fully do that.

What the parser does with empty labels is basically unspecified behaviour, so a unittest over there doesn't prevent someone from changing that behaviour, changing the unittest and breaking the actual feature we care about.

**Comment by krasi-georgiev** [comment]
@brian-brazil yes that makes total sense we need to cover that. 
I added an additional test to make sure the parser doesn't strip empty labels with an additional comment to explain why we need this behaviour.
If you still think this is insufficient  can you give me an example of a possible scenario how this bug can be reintroduced.

**Comment by brian-brazil** [comment]
> My main goal is to prevent reintroducing the bug , but keep the tests as simple as possible to avoid unnecessary refactoring.

This style of testing actually makes refactoring harder, as you need to rewrite the tests if the structure around mutateSamples changes. Just like other tests in this file I would like to test from scrape data all the way to appending, as that is what we really care about.

**Comment by krasi-georgiev** [comment]
yep , you are right here. I refactored it and really happy how it turned out.

**Comment by krasi-georgiev** [comment]
@fabxc I have refactored the TestScrapeLoopAppend to include e2e ingestion path.
This would make it more difficult to maintain and most likely would require more frequent refactoring, but if @brian-brazil thinks it is important I can back off on this one.

I am thinking that better names for the  tests case labels could be `scrapeLabels` , `discoveryLabels.` 
it shows the origin of the labels which makes the test a bit easy to understand.

File: scrape/scrape_test.go [hunk 4]
```diff
@@ -660,42 +661,109 @@ func TestScrapeLoopRunCreatesStaleMarkersOnParseFailure(t *testing.T) {
...
+		expected        []string
+	}{
+		{
+			// When "honor_labels" is not set
+			// label name collision is handler by adding a prefix.
+			title:           "Label name collision",
+			honorLabels:     false,
+			scrapeLabels:    `metric{n="1"} 0`,
+			discoveryLabels: []string{"n", "2"},
+			expected:        []string{"metric", "0", "exported_n", "1", "n", "2"},
```
Comment (fabxc) [inline scrape/scrape_test.go]: Mh, so this encodes the metric name in the first entry, then a value in the 2nd and then label pairs in the subsequent one? Seems rather convoluted and I had to read the test code to check what the test case actually expresses.

Maybe split it up into dedicated fields?

**Comment by fabxc** [comment]
Seems generally fine to me aside from that one comment.

File: scrape/scrape_test.go (hunk 4 above)
Comment (krasi-georgiev) [reply scrape/scrape_test.go]: now when I see it changed it looks so logical and so simple , thanks for the :1st_place_medal:  comments

**Comment by fabxc** [comment]
LGTM. @brian-brazil  for final approval

File: scrape/scrape_test.go [hunk 5]
```diff
@@ -660,42 +660,100 @@ func TestScrapeLoopRunCreatesStaleMarkersOnParseFailure(t *testing.T) {
...
+		expected := []sample{
+			{
+				metric: test.expLset,
+				t:      timestamp.FromTime(now),
+				v:      test.expValue,
+			},
+		}
+
+		// The special case with the NaN is that we only check if the result reports that the value is NaN.
+		if math.Float64bits(app.result[0].v) > 0 {
```
Comment (brian-brazil) [inline scrape/scrape_test.go]: This seems more complicated than it needs to be. Why not check that the value for all cases matches what's expected, using math.Float64bits?

**Comment by brian-brazil** [comment]
Can you squash down your changes?

File: scrape/scrape_test.go (hunk 5 above)
Comment (krasi-georgiev) [reply scrape/scrape_test.go]: the updated version is much cleaner.

**Comment by krasi-georgiev** [comment]
@brian-brazil ping

**Comment by brian-brazil** [comment]
Thanks!
